lib/receptor_controller/client/directive_non_blocking.rb

```diff
@@ -0,0 +1,144 @@
+require "concurrent"
+require "receptor_controller/client/directive"
+
+module ReceptorController
+  # Non-blocking directive for requests through POST /job
+  # Directive's call returns either message ID or nil
+  #
+  # Callback blocks can be specified for handling responses
+  # @example:
+  #         receiver = <object with methods below>
+  #         directive
+  #           .on_success do |msg_id, response|
+  #             receiver.process_response(msg_id, response)
+  #           end
+  #           .on_error do |msg_id, code, response|
+  #             receiver.process_error(msg_id, code, response)
+  #           end
+  #           .on_timeout do |msg_id|
+  #             receiver.process_timeout(msg_id)
+  #           end
+  #           .on_eof do |msg_id|
+  #             receiver.process_eof(msg_id)
+  #           end
+  #           .on_eof do |msg_id|
+  #             logger.debug("[#{msg_id}] EOF message received")
+  #           end
+  #
+  #         directive.call
+  class Client::DirectiveNonBlocking < Client::Directive
```

Comment:
(just putting my thoughts here for next steps)

This class will be re-worked a lot when we switch to redis backed storage for `msg_id`'s. Currently relies on an in-memory hash has that you pass functions as callbacks, this will get moved to a worker that just calls the correct function based on metadata set in the redis message. 

ie if we post a job to the platform receptor controller for source id 42 we get id xyz123 which is an availablity check, so we store: 
`redis.set(:xyz123, {id: 42, action: availability_check}.to_json)`
then, when the worker receives a message it will check to see if redis has any info about said message: `redis.get(:xyz123)` which returns `{id: 42, action: availability_check}`. 

From there it will look at the body of the message, and udpate source id 1 (and endpoint) with the availability status returned.

Reply:
just a note for redis-backed storage. As we'll need to store multiple message IDs under one refresh_state UID (request + subrequests), we'll need to solve concurrency.
https://redis.io/topics/transactions - Redis WATCH seems to be the solution